### DISCUSSION TITLE
Early throttling observed in cluster run with multiple instances of application

### DISCUSSION THREAD

**Discussion by techdeba901** [vladimir-bukhtoyarov/bucket4j#201] [General]
Hi Vladimir,

I am trying to run bucket4j(with coherence) with intervally aligned bucket configuration where tokens are fetched in batch. Batch value is calculated based on incoming message rate(thread calculating this value also aligned with 
second). This is running in cluster environment where 20+ instances are running and processing message with same key. 

Bucket configuration 
rate: 60k TPS

Observation:
At 30k TPS for this message key, throttling starts ~@1.5k. It is not clear why this throttling is happening at such a low rate compared to configured rate in the bucket. 

When I used refill strategy as greedy it's behavior improves a little and throttling starts @43k TPS for that key.

Attached the sample code I am using for this run. 

Could you please help in understanding what possibly go wrong here?

Thanks in advance.

```java
//Bucket configurations
public BucketConfiguration buildBucket(int rate, int durationInSec) {
		List<Bandwidth> bandwidths = new ArrayList<>();
		Instant instant = Instant.now().truncatedTo(ChronoUnit.SECONDS).plus(1, ChronoUnit.SECONDS);
		bandwidths.add(Bandwidth.classic(rate,
				Refill.intervallyAligned(rate, Duration.ofSeconds(durationInSec), instant, true)));
		ConfigurationBuilder bucketConfiguration = Bucket4j.configurationBuilder();
		bandwidths.stream().forEach(bw -> bucketConfiguration.addLimit(bw));
		return bucketConfiguration.build();
	}

	
//Cache Initialization	
private ProxyManager<String> bucketProxy;

	@PostConstruct
	private void intializeRateLimitService() {
		NamedCache cache = CacheFactory.getCache("testMap");
		cache.addIndex(new DeserializationAccelerator(IdentityExtractor.INSTANCE()), false, null);
		this.bucketProxy = Bucket4j.extension(Coherence.class).proxyManagerForCache(cache);
	}	


//Token fetch logic
public synchronized Boolean fetchToken(String bucketKey, Supplier<BucketConfiguration> configurationSupplier,
			int batch) {
		Long probe = 0L;

		// Try to access token from local map first
		AtomicInteger token = cacheMap.get(bucketKey);
		if (token != null && token.get() > 0) {
			if (token.get() == 1) {
				cacheMap.get(bucketKey).getAndSet(0);
			} else {
				cacheMap.get(bucketKey).decrementAndGet();
			}
			return true;
		}

		// Lets try to fetch token using bucket4j sync call with Coherence
		boolean tokenObtained = false;
		Bucket bucket = bucketProxy.getProxy(bucketKey, configurationSupplier);
		probe = bucket.tryConsumeAsMuchAsPossible(batch);

		if (probe > 0) {
			if (cacheMap.containsKey(bucketKey)) {
				cacheMap.get(bucketKey).addAndGet(probe.intValue() - 1);
			} else {
				cacheMap.put(bucketKey, new AtomicInteger(1));
			}
			tokenObtained = true;
		}
		return tokenObtained;
	}	
	
libraries used
--------------	
	
    implementation "com.oracle.coherence.ce:coherence:21.06.2"
    implementation "com.github.vladimir-bukhtoyarov:bucket4j-coherence:6.4.1"
    implementation "com.github.vladimir-bukhtoyarov:bucket4j-core:6.4.1"	
```

**Comment by vladimir-bukhtoyarov** [comment]
Off-top, I see that you use the highly loaded bucket.  Today I have been released the 7.0.0 version that contains significant enhancement for such case #90 . I am very surprised to know that Coherence is able to process 30K TPS on same key, I did not use to believe in such numbers before. You can significantly reduce pressure to the Coherence cluster by migrating to 7.0.0 and migrating the code a bit:
```java
        ...
        @PostConstruct
	private void intializeRateLimitService() {
		NamedCache cache = CacheFactory.getCache("testMap");
		cache.addIndex(new DeserializationAccelerator(IdentityExtractor.INSTANCE()), false, null);
		this.bucketProxy = new CoherenceManager(cache);
		
                // optimized bucket is not lightweith so it can be stored somewhere for reuse
                this.bucket = bucketProxy.builder()
                                          .withOptimization(Optimizations.batching())
                                          .build(bucketKey, configurationSupplier);
	}
```

**Reply by techdeba901** [reply]
Thanks for your input.

Just for the clarification, the run I attempted earlier was invoking tryConsumeAsMuchAsPossible with a batch count derived from incoming message rate. Hence the total no of coherence cache hits were much lesser than the total TPS at which messages were injected. With 30K TPS the coherence cache hits were around 2.5k per second. But as reported at 30k TPS suddenly throttling started even though the bucket was configured with 60k TPS.

Regarding bucket4j release 7.0.0, I have given an attempt today. Please find details below.

Sample code used:

public Boolean isTokenAvailable(String bucketIdentifier, Supplier<BucketConfiguration> configurationSupplier ) {
		boolean probe = false;
		
		// Lets try to fetch token using bucket4j sync call with Coherence
		boolean tokenObtained = false;
		try {
			Bucket bucket = getBucket(bucketIdentifier, configurationSupplier);
			probe = bucket.tryConsume(1);
			if (probe) {
				tokenObtained = true;
				logger.debug("Token fetched for  " + bucketIdentifier + " + " + probe);
			}
		}catch (Exception e) {
			logger.catching(Level.ERROR, e);
		}
		return tokenObtained;
	}

	Map<String, Bucket> bucketMap = new HashMap<>();

	public synchronized Bucket getBucket(String bucketIdentifier, Supplier<BucketConfiguration> configurationSupplier) {
		if (!bucketMap.containsKey(bucketIdentifier)) {
			Bucket bucket = getBucketProxy().builder().withOptimization(Optimizations.batching())
					.build(bucketIdentifier, configurationSupplier);
			bucketMap.put(bucketIdentifier, bucket);
		}
		return bucketMap.get(bucketIdentifier);
	}

Version:
bucket4j-coherence:7.0.0
bucket4j-core : 7.0.0

Observation:
I did  a small run with 100 TPS for 60 seconds and I can see total coherence cache hit count as 6k at the end of the run. Meaning for every message cache is accessed (ran it few times to confirm this behavior).

Questions:
1. I was expecting fewer cache hits, since with batching, coherence will be accessed in intervals ? Is this a correct understanding ?
2. If yes, how the interval of syncing across cluster is decided when bucket is configured with rate of X no of messages per second  ?
3. To hit batching, do I need to trigger very high traffic ? Is there any specific rate value ?
4. Not sure, whether I have used the batching API of bucket4j correctly. Please correct me if I have done anything wrong.
5. Have plans to try with async API as well once the numbers with sync is obtained.

**Reply by vladimir-bukhtoyarov** [reply]
@techdeba901

1.  In the provided example you create a bucket per request, and batching optimization does not work, you need to store once created bucket. Please pay attention to my comment and [this javadocs](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L52)
> // optimized bucket is not lightweith so it can be stored somewhere for reuse

2. There is no background syncing, synchronization is always performed in the scope of user interaction, but it is not defined in which concrete. Instead of issuing requests to the Coherence cluster one by one, the batching combines independent requests in a single batch(that looks for Coherence as single invocation of entry processor) and does not allow to execution of the next batch until the previous batch is fully executed. My quick tests show that batching optimization allows processing millions of throttling operations on the same key per second.

3. Answered in the item 1

4. Follow this instructions for async API https://bucket4j.com/7.0.0/toc.html#asynchronous-api

**Reply by vladimir-bukhtoyarov** [reply]
Also this code can be used as example https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-examples/src/test/java/example/distributed/optimizers/batching/HazelcastWithBatchingPerformanceExample.java

**Reply by Buro45** [reply]
@vladimir-bukhtoyarov 
In this case the bucket is created once in the following function and saved in a MAP (bucketMap which is member of the class) against the bucket identifier. From the incoming message bucket identifier is derived and used as key to get the Bucket object from the MAP. 

When the 1st request comes bucket will be built for the first time and stored in the map. And then from 2nd request onwards bucket object will be retrieved from the map using bucket identifier.

```
public synchronized Bucket getBucket(String bucketIdentifier, Supplier<BucketConfiguration> configurationSupplier) {
    // This if check passed only one time
    if (!bucketMap.containsKey(bucketIdentifier)) {
		Bucket bucket = getBucketProxy().builder().withOptimization(Optimizations.batching())
				.build(bucketIdentifier, configurationSupplier);
		bucketMap.put(bucketIdentifier, bucket);
	}
	return bucketMap.get(bucketIdentifier);
}
```

Is this not correct way of reusing the bucket object  ?

**Reply by vladimir-bukhtoyarov** [reply]
@Buro45 I missed that bucket is cached. It is the right way, I think that independent requests are not grouped into batches because of low request rate. In order to see the advantages of batching try to do the following:
* Make requests from several threads in parallel
* Make requests with the more frequent rate in order to the bring request intersection.

**Reply by techdeba901** [reply]
@vladimir-bukhtoyarov 
I run with synchronous approach with greedy as refill strategy and run traffic in parallel. I had configured rate limit 60K TPS per key .

**Synchronous approach:**
At 40k TPS for this message key, throttling started to show up with 6-8 TPS.

**code snippet** :

Bucket bucket = getBucket(bucketIdentifier, configurationSupplier);
probe = bucket.tryConsume(1);


public synchronized Bucket getBucket(String bucketIdentifier, Supplier<BucketConfiguration> configurationSupplier) {
		if (!bucketMap.containsKey(bucketIdentifier)) {
			Bucket bucket = getBucketProxy().builder().withOptimization(Optimizations.batching())
					.build(bucketIdentifier, configurationSupplier);
			bucketMap.put(bucketIdentifier, bucket);
		}
		return bucketMap.get(bucketIdentifier);
	}


**Asynchronous approach:**
When I run asynchronous approach with greedy as refill strategy and run traffic in parallel. I had configured rate limit 60K TPS per key.

At 16k TPS for this message key, I am observing memory leak. So traffic is proceeding further.

**code snippet** :

AsyncBucketProxy asyncBucket = getBucket(bucketIdentifier, configurationSupplier);
CompletableFuture<Boolean> isSuccess = asyncBucket.tryConsume(1);

public synchronized AsyncBucketProxy getBucket(String bucketIdentifier, Supplier<BucketConfiguration> configurationSupplier) {
		if (!bucketMap.containsKey(bucketIdentifier)) {
			AsyncBucketProxy asyncBucket = getBucketProxy().builder().withOptimization(Optimizations.batching())
					.build(bucketIdentifier, configurationSupplier.get());
			bucketMap.put(bucketIdentifier, asyncBucket);
		}
		return bucketMap.get(bucketIdentifier);
	}

**Reply by vladimir-bukhtoyarov** [reply]
Regarding asynchronous approach - you did not get a memory leak, it was just OutOfMemoryError. It is not surprising when tasks are issued  with rate that system is not ready to process. Look to my code, I used Semaphore to limit the number of incomplete tasks https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-examples/src/test/java/example/distributed/optimizers/batching/HazelcastWithBatchingPerformanceExample.java#L149

Regarding reproducing steps in synchronous mode: did you reproduce early throttling from a single thread?

**Comment by vladimir-bukhtoyarov** [comment]
@techdeba901 According to original question: 
I want to start from following line of code
```java
bandwidths.add(Bandwidth.classic(rate, Refill.intervallyAligned(rate, Duration.ofSeconds(durationInSec), instant, true)));
```
You invoke ```intervallyAligned``` with  *useAdaptiveInitialTokens=true*. Was the javadocs clear for you? Do you still to view the problem if *useAdaptiveInitialTokens=false*?

**Reply by vladimir-bukhtoyarov** [reply]
Could you provide a self-contained example that can be executed on the local machine?

**Reply by Buro45** [reply]
@vladimir-bukhtoyarov 
Possibly when we are running the batching approach with greedy as Refill strategy, and with burst in traffic, the periodic refill is falling short of the continuous batch request leading to throttling @41 - 42K TPS, when bucket is configured to support 60K TPS. 

Here, Refill strategy with intervallyAligned should have worked well, since at one shot the entire 60K tokens are allocated in the bucket. But it didn't work for some unknow reason. 

Is this a correct understanding ?

**Reply by vladimir-bukhtoyarov** [reply]
@Buro45 batching is orthogonal to the early throttling that you observed. I just noticed that you faced with huge request rate on the same key and proposed the new functionality that was specially created to reduce the pressure to the Coherence cluster. Maybe it was the wrong advance because I did it without knowledge of numbers that you pass to ```consumeAsMuchAsPossible```.

Let's get back to original question Could you provide an example that I can run on my local machine?

**Reply by techdeba901** [reply]
@vladimir-bukhtoyarov This program I am working on. I will share with you soon.

**Reply by vladimir-bukhtoyarov** [reply]
I was unable to reproduce the problem.

@techdeba901 please modify this class https://github.com/vladimir-bukhtoyarov/bucket4j/blob/201_early_throttling_investigation/bucket4j-coherence/src/test/java/EarlyThrottlingInvestigationTest.java in a way that allows seeing the problem.

**Reply by techdeba901** [reply]
I want to share you the piece code which is running  with different instances(pod count 20+) in the cluster(k8s). This is not a self contained program. Here, refill strategy is configured as intervally aligned with useAdaptiveInitialTokens=false and batch count is hard coded as 20.
In our actual program this batch is derived from the incoming message rate.
I tried one more run with my application (running in 20+ PODs)  , key configured with 30K TPS, traffic generated 25K TPS
   - where each application is running a scheduler thread per second (with second boundary aligned)
   - In this thread, every 100ms, I am invoking estimateAbilityToConsume(1) for 1 buckets (that is used in the program) to check availability of tokens in the bucket
   - Around 17K TPs, Token got exhausted, I obsereved the following logs in one of the instance
   - {"message":"The scheduler is out of sync for key:test_bucket_1 with ms:399,  "messageTimestamp":"2021-12-20T08:03:15.601+0000",}
     {"message":"The scheduler is out of sync for key:test_bucket_1 with ms:298, "messageTimestamp":"2021-12-20T08:03:15.702+0000"}
	 {"message":"The scheduler is out of sync for key:test_bucket_1 with ms:198, "messageTimestamp":"2021-12-20T08:03:15.802+0000"}
     {"message":"The scheduler is out of sync for key:test_bucket_1 with ms:98, "messageTimestamp":"2021-12-20T08:03:15.902+0000"}
	
   - From the above logs, looks like there is no alignment between the consuming application and the time when bucket is refilled.
   
Code is attached for your reference.

[AppController.txt](https://github.com/vladimir-bukhtoyarov/bucket4j/files/7761827/AppController.txt)
[Bucket4jService.txt](https://github.com/vladimir-bukhtoyarov/bucket4j/files/7761829/Bucket4jService.txt)


- Trying to understand how this is possible.  Any thoughts ?

**Reply by vladimir-bukhtoyarov** [reply]
@techdeba901 

Please recheck the code of Bucket4jService. It can not be compiled because variable is declared twice tokenObtained.

Also the code contains logical error
![image](https://user-images.githubusercontent.com/6761830/146788881-f6ffaad6-dfa6-42ca-aa44-3404c09430f3.png)

**Reply by techdeba901** [reply]
Sorry. Updated the corrected file.

**Reply by vladimir-bukhtoyarov** [reply]
Also the name isTokenAvailable is not clear for me. Name of method tells that method will not modify something, but behind it modifies the state of bucket by calling **tryConsumeAsMuchAsPossible**. Why is not used [estimateAbilityToConsume]( https://bucket4j.com/7.0.0/toc.html#estimateabilitytoconsume) that does not modify the state of bucket?

**Reply by vladimir-bukhtoyarov** [reply]
@techdeba901 

> Updated the corrected file.

But code still containing ```cacheMap.put(bucketKey, new AtomicInteger(1));``` instead of ```cacheMap.put(bucketKey, new AtomicInteger(tokenConsumed.intValue()  - 1));```

**Reply by techdeba901** [reply]
you are right. Updated the files.

**Reply by vladimir-bukhtoyarov** [reply]
@techdeba901 I still see new AtomicInteger(1)) at the attached file.

Anyway it is time to use Verbose API, to better understand what is going on:
```java
public synchronized Boolean isTokenAvailable(String bucketKey,
                                                 Supplier<BucketConfiguration> configurationSupplier) {
        Long tokenConsumed = 0L;
        // Try to access token from local map first
        AtomicInteger token = cacheMap.get(bucketKey);
        if (token != null && token.get() > 0) {
            if (token.get() == 1) {
                cacheMap.get(bucketKey).getAndSet(0);
            } else {
                cacheMap.get(bucketKey).decrementAndGet();
            }
            return true;
        }
        // Lets try to fetch token using bucket4j sync call with Coherence
        boolean tokenObtained = false;
        try {
            Bucket bucket = bucketProxy.getProxy(bucketKey, configurationSupplier);
            /*
             * Here the token count is passed as 20. But in the original program this token
             * count is decided based on incoming message rate
             */
            Date myTime = new Date();
            VerboseResult<Long> tokenConsumedResult = bucket.asVerbose().tryConsumeAsMuchAsPossible(20);
            tokenConsumed = tokenConsumedResult.getValue();

            if (tokenConsumed > 0) {
                if (cacheMap.containsKey(bucketKey)) {
                    cacheMap.get(bucketKey).addAndGet(tokenConsumed.intValue() - 1);
                } else {
                    cacheMap.put(bucketKey, new AtomicInteger(tokenConsumed.intValue() - 1));
                }
                tokenObtained = true;
            } else {
                Date operationTimestampThatWasActualOnServer = new Date(tokenConsumedResult.getOperationTimeNanos() / 1_000_000);
                long nanosToRefillToOriginalState = tokenConsumedResult.getDiagnostics().calculateFullRefillingTime();
                logger.warn("Tokens was not consumed at server-timestamp " + operationTimestampThatWasActualOnServer + ". It was needed to wait " + nanosToRefillToOriginalState + 
                        " nanoseconds before full bucket refill, my timestamp was " + myTime);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return tokenObtained;
    }
```
Now you are able to see timestamp on node that holds the state of bucket and timestamp on node that got "0" as result of tryConsumeAsMuchAsPossible operation. Also message in log shows the time in nanoseconds that need to wait before bucket will be fully refilled to original state.

**Comment by techdeba901** [comment]
we observed from the source code of version 6 that async implementation of bucket depends on the CommonForkJoinPool which can be very less. Is there a way in bucket4j to pass your own executor pool. If not, is there a way to define parallelism for the async methods

**Reply by vladimir-bukhtoyarov** [reply]
@techdeba901 Bucket4j in the background does not use CommonForkJoinPool as well as any other thread pool in all versions. 
However, it can be used on behalf of you when you use methods like thenApply/whenComplete and similar without passing own executor as an explicit parameter. Javadocs for each method of async API contains appropriate notice about that, for example https://github.com/vladimir-bukhtoyarov/bucket4j/blob/6.4/bucket4j-core/src/main/java/io/github/bucket4j/AsyncBucket.java#L53

I have debugged Coherence by itself in order to find where it executes aynch operations result callbacks, it always happens inside own Coherence threads
![image](https://user-images.githubusercontent.com/6761830/147359482-46736cb6-d33d-418e-8ad5-a9cc9e5341ce.png)

So if you see an usage of CommonForkJoinPool then double-check what you do with CompleteableFutures.

**Comment by techdeba901** [comment]
Thanks Vladimir for your inputs.

Tokens are added in 100ms when configured rate is 1000 per second (i.e. 100 tokens in every 100ms) when refill strategy is greedy.
But if consumption of tokens is more than 100 per 100ms will tokens be added more greedily(more frequently).

In tryConsumeAndReturnRemaining, when we use this method it return following 3 variables(remainingTokens , nanosToWaitForRefill, nanosToWaitForReset) in version 7.1.0

What is the difference between nanosToWaitForRefill & nanosToWaitForReset?
Assume the tokens are exhausted @800ms within a second boundary.
In that case, does nanosToWaitForReset will be 200ms? 
Could you please provide your input?

**Reply by vladimir-bukhtoyarov** [reply]
**nanosToWaitForRefill** is zero if tokens were consumed, otherwise it describes the time that needs to wait before the requested amount of tokens can be consumed.

**nanosToWaitForReset** describes the time required to refill all tokens up to max capacity.

**Reply by techdeba901** [reply]
What will happen when all tokens will be consumed by half of 1 second(refill config is 1000/s & refill strategy is greedy ) and application ask for 100 tokens?

In the above case, nanosToWaitForRefill value will be 500ms(which is start of the next second)?

**Reply by vladimir-bukhtoyarov** [reply]
It will be ```ConsumptionProbe{consumed=false, remainingTokens=0, nanosToWaitForRefill=100000000, nanosToWaitForReset=1000000000}```

Answer to this and similar questions you can find just by modeling with mock timer:
```java
public static void main(String[] args) {
        AtomicLong currentTimeMillis = new AtomicLong(System.currentTimeMillis());
        TimeMeter mockClock = new TimeMeter() {
            @Override
            public long currentTimeNanos() {
                return 1_000_000 * currentTimeMillis.get();
            }

            @Override
            public boolean isWallClockBased() {
                return false;
            }
        };

        Bucket bucket = Bucket.builder()
                .withCustomTimePrecision(mockClock)
                .addLimit(Bandwidth.classic(1000, Refill.greedy(1000, Duration.ofSeconds(1))))
                .build();

        // emulate 500 milliseconds delay
        currentTimeMillis.addAndGet(500);

        System.out.println(bucket.tryConsumeAndReturnRemaining(1000));
        System.out.println(bucket.tryConsumeAndReturnRemaining(100));
    }
```

**Reply by techdeba901** [reply]
I am trying understand, despite of the greedy strategy, shouldn't the total token consumption be restricted to 1000/sec in this case. Considering that next refill shouldn't happen before 500ms right ? if we allow the next 100 refill in 100ms that would mean in 600ms we are allowing 1100 tokens to be consumed right ?

**Reply by vladimir-bukhtoyarov** [reply]
The next refill is allowed to happen after 1ms, because refill rate _1000 tokens/1 second_ fully equals _1 token/1 millisecond_.
The consumption rate is not a the part of the TokenBucket algorithm, if the token is allowed to be refilled according to the refill setting then it will be refilled independently of the current consumption rate.

**Reply by techdeba901** [reply]
In Greedy, is there a way to define refill interval instead of keeping it 1ms without using more than one BW? For example, I want 5000/sec but refill should happen at every 100 ms

**Reply by vladimir-bukhtoyarov** [reply]
@techdeba901 you should switch to interval refill in such case. For example:
```java
Bandwidth.classic(5000, Refill.intervally(500, Duration.ofMillis(100)));
```

**Reply by techdeba901** [reply]
Considering above example, i.e. refill 500 per 100ms, at any point of time if token get exhausted then next refill will happen at the nearest 100ms  timestamp. Hence, nanosToWaitForRefill will represent the time remaining to the next 100ms boundary when next refill will happen. Is this a correct understanding?

**Reply by vladimir-bukhtoyarov** [reply]
Yes, it is how the intervally refill works.
